### ISSUE TITLE
duplicated Tasks - identical text, notes, and ID. Different challenge element: empty object versus null

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#4470]
ping @lefnire @paglias 

In the Tavern, a user reported that their Habits had been duplicated: _"Hi, can someone tell me if my Habits will reset themselves daily? I only have four that I'm working on, but they are listed now twice in my Habit sections, so I didn't know if it was a bug or if that's what I should expect?"_

The user does not have any Dailies, To-Dos, or Rewards so I can't tell if this bug affects only Habits or all tasks.

Their data displaying this problem is below. It's identical to what was in the database except that I have replaced all text and notes with placeholders like "Text 1", etc. Where the same piece of text appeared twice, I replaced it with the same placeholder.

You can see the following things:
- All but one of their Habits was duplicated ("Text 4" had no duplicate).
- Each duplicated Habit had identical text, note, and ID.
- Each duplicated Habit had **different challenge elements**: one copy had an empty object `{ }` and the other copy had `null`.

```
{
    "_id" : "de860fb0-159d-4ed5-9a32-9484c946b80a",
    "habits" : [
        {
            "text" : "Text 1",
            "challenge" : {

            },
            "attribute" : "str",
            "priority" : 2,
            "value" : 0,
            "tags" : {

            },
            "notes" : "Note 1",
            "dateCreated" : ISODate("2015-01-02T18:24:04.478Z"),
            "id" : "8c6d6092-656c-49fa-b8e2-7d9854532e10",
            "down" : true,
            "up" : true,
            "history" : [ ],
            "type" : "habit"
        },
        {
            "text" : "Text 2",
            "challenge" : {

            },
            "attribute" : "str",
            "priority" : 1.5,
            "value" : 0,
            "tags" : {

            },
            "notes" : "Note 2",
            "dateCreated" : ISODate("2015-01-02T18:20:03.030Z"),
            "id" : "039ef9b3-cbea-4674-9270-f3a773be324d",
            "down" : true,
            "up" : true,
            "history" : [ ],
            "type" : "habit"
        },
        {
            "text" : "Text 3",
            "challenge" : {

            },
            "attribute" : "str",
            "priority" : 2,
            "value" : 0,
            "tags" : {

            },
            "notes" : "Note 3",
            "dateCreated" : ISODate("2015-01-02T18:21:54.688Z"),
            "id" : "d4e662cd-2f9b-480e-9cef-50dccc32d5f4",
            "down" : true,
            "up" : true,
            "history" : [ ],
            "type" : "habit"
        },
        {
            "text" : "Text 4",
            "challenge" : {

            },
            "attribute" : "str",
            "priority" : 1,
            "value" : 0,
            "tags" : {

            },
            "notes" : "",
            "dateCreated" : ISODate("2015-01-02T19:05:31.320Z"),
            "id" : "fba1e045-5921-4dfb-b0db-6e00cb240746",
            "down" : true,
            "up" : true,
            "history" : [ ],
            "type" : "habit"
        },
        {
            "text" : "Text 5",
            "challenge" : {

            },
            "attribute" : "str",
            "priority" : 2,
            "value" : 0,
            "tags" : {

            },
            "notes" : "Note 5",
            "dateCreated" : ISODate("2015-01-02T18:46:08.374Z"),
            "id" : "13fc65ed-8440-43b5-9b1f-584d3d4511ee",
            "down" : false,
            "up" : true,
            "history" : [ ],
            "type" : "habit"
        },
        {
            "text" : "Text 5",
            "challenge" : null,
            "attribute" : "str",
            "priority" : 2,
            "value" : 0,
            "tags" : {

            },
            "notes" : "Note 5",
            "dateCreated" : ISODate("2015-01-02T18:46:08.374Z"),
            "id" : "13fc65ed-8440-43b5-9b1f-584d3d4511ee",
            "down" : false,
            "up" : true,
            "history" : [ ],
            "type" : "habit"
        },
        {
            "text" : "Text 1",
            "challenge" : null,
            "attribute" : "str",
            "priority" : 2,
            "value" : 0,
            "tags" : {

            },
            "notes" : "Note 1",
            "dateCreated" : ISODate("2015-01-02T18:24:04.478Z"),
            "id" : "8c6d6092-656c-49fa-b8e2-7d9854532e10",
            "down" : true,
            "up" : true,
            "history" : [ ],
            "type" : "habit"
        },
        {
            "text" : "Text 3",
            "challenge" : null,
            "attribute" : "str",
            "priority" : 2,
            "value" : 0,
            "tags" : {

            },
            "notes" : "Note 3",
            "dateCreated" : ISODate("2015-01-02T18:21:54.688Z"),
            "id" : "d4e662cd-2f9b-480e-9cef-50dccc32d5f4",
            "down" : true,
            "up" : true,
            "history" : [ ],
            "type" : "habit"
        },
        {
            "text" : "Text 2",
            "challenge" : null,
            "attribute" : "str",
            "priority" : 1.5,
            "value" : 0,
            "tags" : {

            },
            "notes" : "Note 2",
            "dateCreated" : ISODate("2015-01-02T18:20:03.030Z"),
            "id" : "039ef9b3-cbea-4674-9270-f3a773be324d",
            "down" : true,
            "up" : true,
            "history" : [ ],
            "type" : "habit"
        }
    ]
}
```

I have since removed the duplicates from the user's account.

This might be related to https://github.com/HabitRPG/habitrpg/issues/2456 but that issue is pretty old so it might be a separate problem.

**Comment by jbbreck** [comment]
This happened to my account too. It repeated a bunch of my habits and checked some dailies without my click. Here's my user ID: ee41f326-c4df-44e4-a966-d1962877f5b3

Edit: It also took away some of my health.

**Comment by Alys** [comment]
@jbbreck I can investigate your duplicated tasks in about 24 hours (Saturday morning for me) if no one else gets to it first.

It isn't likely that the duplicated task bug removed health because those two parts of the site are very separate. Instead, something else would have removed it. If you left Dailies incomplete yesterday, then health would have been subtracted when you first used HabitRPG this morning. If you clicked on any negative Habits, then health would have been subtracted, and if you did that in another browser tab or on a different device, then you would have received the notification about missing health in both locations (but the health loss itself would have been performed only once). You can learn about the causes of health loss and how to regain health from the wiki's [Health Points](http://habitrpg.wikia.com/wiki/Health_Points) page.

**Comment by jbbreck** [comment]
Thanks Alice! I appreciate your help.

-Jacob

**Comment by Alys** [comment]
@jbbreck Your duplicated tasks have been removed. Let us know if they come back! 

Your Dailies being ticked without you doing it yourself is odd but I can't see anything that might have caused that. I am hoping it was just a once-off glitch. Please also tell us if it happens again.

**Comment by LittleELK26** [comment]
I have this issue too, all 3 types of tasks have been duplicated, although only some of them are showing duplicates.  My used ID is: d288ea32-3b06-41a8-988a-fe96af9dcbde

**Comment by Alys** [comment]
@LittleELK26 One of your To-Dos has been duplicated so that there are now six copies of it (all of them marked as completed) but I can't see any duplication in the other two task types. Are you seeing something different? Can you please give me more information about it?

**Comment by LittleELK26** [comment]
Hi @Alys Sorry for the delay- I was trying to manually remove the duplicates and, while it didn't seem to register for a while, eventually I managed to clean them all out myself.  This problem has intermittently repeated itself (though not at the moment) but is usually only repeating one or two list items once or twice rather than the dozen+ it was doing last week.

**Comment by weirdpancake** [comment]
My to-do  list tasks are more than duplicated. A few of them repeat at least 10 times. I can't delete them, and if I try to check off any of my other to-do tasks, I lose health, money, and experience. :(

**Comment by deilann** [comment]
You need to supply your User ID or else no one can investigate your account.

Furthermore, have you tried deleting them from the bottom up? As stated,
that works for a lot of users.

**Comment by weirdpancake** [comment]
Oh, sorry. I forgot that github isn't connected to my habitrpg account.

My account is connected to my facebook account, so I'm not sure what
information to give you.

**Comment by weirdpancake** [comment]
I think this might be my designated user ID:

9f4859fe-beb5-465d-af68-cd7edb8f574a

**Comment by deilann** [comment]
Yup, looks like it! Someone with database should be able to help soon.

However, did you try deleting starting with the bottom task?

**Comment by weirdpancake** [comment]
Yeah, I just did that. It worked, but I'm still having the issue where if I
try to check off a to-do task it takes away life, gold, exp, and mp.

**Comment by deilann** [comment]
Do you have the Chrome extension installed?

**Comment by weirdpancake** [comment]
Ok. Scratch that. It's working now. I only posted the bug because it's been
doing it all morning, but now it stopped. If it starts happening again I'll
post the information.

**Comment by CCapra** [comment]
I from #5189, my user id is e2328037-6966-4b36-9d0f-81326c3ed38c - in case that helps.
I was able to delete from the bottom up, but the status of my tasks was a little wonky. 

Does anyone have clues to what's causing it? i.e. things I could avoid doing to keep it from happening again? and btw, I'm brand new to Hrpg, but I really like it - so I'm hoping I can keep using it.

**Comment by crookedneighbor** [comment]
I believe this was fixed by #4838

**Comment by crookedneighbor** [comment]
Re-open if it continues to happen.
